cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -309,6 +313,202 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
   return results;
 }
 
+namespace {
+
+/// Workhorse for streaming async implementation of `GetFileInfo`
+/// (`GetFileInfoGenerator`).
+///
+/// There are two variants of async discovery functions suported:
+/// 1. `DiscoverPartitionFiles`, which parallelizes traversal of individual directories
+///    so that each directory results are yielded as a separate `FileInfoGenerator` via
+///    an underlying `DiscoveryImplIterator`, which delivers items in chunks (default size
+///    is `kBatchSize == 1K` items).
```

Comment:
I don't think delivering items in fixed-size chunks is a worthwhile goal, on the countrary.
The purpose of an async API such as `GetFileInfoGenerator` is to deliver results to the consumer as soon as they arrive. So when you get the contents of a 5-file directory, emit them ASAP as a single chunk. When you get the contents of a 999999-file directory, emit them ASAP as a single chunk.

The consumer may want to normalize batch size if that makes sense for them, but it might not (for example if you simply print files on stdout, why would you enforce a batch size?).

This will also simplify the producer implementation :-)

Reply:
> I don't think delivering items in fixed-size chunks is a worthwhile goal, on the contrary.
The purpose of an async API such as GetFileInfoGenerator is to deliver results to the consumer as soon as they arrive. So when you get the contents of a 5-file directory, emit them ASAP as a single chunk. When you get the contents of a 999999-file directory, emit them ASAP as a single chunk.

This is exactly what this impl does. Each directory iterator has its own chunk tracking and flushes the chunks either as they are filled up (to `kBatchSize` entries) or the iterator is exhausted (in this case `current_chunk_.size()` would be less than `kBatchSize`). So I see there no problem at all, although the chunk size should be a configurable option, too.

Also, I just noticed that I always reserve `kBatchSize` elements for a chunk, regardless of how much elements are left unprocessed. In such case the allocation size can optimized by checking against `child_fns_.size() - idx_`. I'll fix that one.

Reply:
If we yield on every new item in the queue, more work will be just wasted on moving back and forth between IO and CPU threads, AFAIU, so internal chunking makes perfect sense. Moreover, it's naturally supported by the signature of `GetFileInfoGenerator`, which is supposed to deliver items in some chunks (though, without indication of batch sizes).

If consumer wants a flattened view (e.g. `AsyncGenerator<FileInfo>`), it would want use more primitives from `async_generator.h` anyway.

Reply:
I'm not really following you. What use is the chunk size here? If you just got 100000 elements from the kernel, while yield every 1000 of them, for example?

Reply:
Getting just directory entries (`linux_dirent`) from linux kernel is extremely fast, even though we use regular `readdir`, which fetches the items one-by-one, and don't use `getdents` syscall directly, which is capable of yielding multiple entries at once and should be even faster.

For example, on my machine, fetching contents of a directory with 10M files completes in less than 2 seconds, which is acceptable, I think.

The most time-consuming part is `stat`-ing individual files, which should be done sequentially by arrow code and there isn't really a _get 100000 items from the kernel_ option, so the batching resides on our part, too.

Reply:
Ah, I had forgotten about the `stat`ing, my bad.

Reply:
I'll extend the comments to clarify these things :)

Reply:
Is there any benefit to running `stat` in parallel?  For example, if we are listing a single directory with 100,000 files.  Would it make sense to kick off 10 thread tasks, each one `stat`ing 10k files?

Reply:
In the current approach, a single directory is processed in a synchronous fashion. I haven't yet tested whether it would be beneficial to further parallelize `stat` calls or not.